src/main/java/com/example/auctionapp/repository/ProductRepository.java

```diff
@@ -0,0 +1,41 @@
+package com.example.auctionapp.repository;
+
+import com.example.auctionapp.model.Product;
+import com.example.auctionapp.model.Subcategory;
+import com.example.auctionapp.model.User;
+import org.springframework.stereotype.Repository;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import java.util.List;
+
+@Repository
+public class ProductRepository extends BaseRepository<Product> {
+
+    @PersistenceContext
+    EntityManager entityManager;
+
+    public List<Product> findRelatedProducts(Long productId, Long subcategoryId) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+
+        CriteriaQuery<Product> q = cb.createQuery(Product.class);
+        Root<Product> resource = q.from(Product.class);
+        q.select(resource);
+        Predicate predicateForSubcategory = cb.equal(resource.get("subcategory").get("id"), subcategoryId);
+        Predicate predicateForProduct = cb.notEqual(resource.get("id"), productId);
+        q.where(
+                cb.and(
+                        predicateForProduct,
+                        predicateForSubcategory
+                )
+        );
+
+        List<Product> result = entityManager.createQuery(q).getResultList();
```

Comment:
This is why pagination should be implemented. There is no limit on how many products will be returned, which will create memory issues. 
For now, please limit the returned products to the maximum number needed client-side (i guess its 4?)

Reply:
Arent products basically biddings? They should have start date and end date. With this query you will return the finished ones as well, i suggest you limit the results that are in the similar time frame as the original product